apps/arweave/src/ar_peers.erl

```diff
@@ -552,59 +641,134 @@ rank_peers(ScoredPeers) ->
 			[],
 			GroupedBySubnet
 		),
-	[Peer || {Peer, _} <- lists:sort(fun({_, S1}, {_, S2}) -> S1 >= S2 end,
-			ScoredSubnetPeers)].
+	[Peer || {Peer, _} <- lists:sort(
+		fun({_, S1}, {_, S2}) -> S1 >= S2 end,
+		ScoredSubnetPeers
+	)].
 
-check_external_peer(Peer) ->
-	IsLoopbackIP = is_loopback_ip(Peer),
+set_ranked_peers(Rating, Peers) ->
+	ets:insert(?MODULE, {{peers, Rating}, lists:sublist(Peers, ?MAX_PEERS)}).
+
+check_peer(Peer) ->
+	check_peer(Peer, not is_loopback_ip(Peer)).
+check_peer(Peer, IsPeerScopeValid) ->
 	IsBlacklisted = lists:member(Peer, ?PEER_PERMANENT_BLACKLIST),
 	IsBanned = ar_blacklist_middleware:is_peer_banned(Peer) == banned,
-	case {IsLoopbackIP, IsBlacklisted, IsBanned} of
-		{true, _, _} ->
-			reject;
-		{_, true, _} ->
-			reject;
-		{_, _, true} ->
-			reject;
-		_ ->
-			ok
+	case IsPeerScopeValid andalso not IsBlacklisted andalso not IsBanned of
+		true ->
+			ok;
+		false ->
+			reject
 	end.
 
-update_rating(Peer, TimeDelta, Size) ->
+update_rating(Peer, IsSuccess) ->
+	update_rating(Peer, undefined, undefined, 1, IsSuccess).
+update_rating(Peer, LatencyMilliseconds, DataSize, Concurrency, false)
+  		when LatencyMilliseconds =/= undefined; DataSize =/= undefined ->
+	%% Don't credit peers for failed requests.
+	update_rating(Peer, undefined, undefined, Concurrency, false);	
+update_rating(Peer, LatencyMilliseconds, DataSize, Concurrency, IsSuccess) ->
 	Performance = get_or_init_performance(Peer),
-	Total = get_total_rating(),
-	#performance{ bytes = Bytes, time = Time,
-			rating = Rating, transfers = N } = Performance,
-	Bytes2 = Bytes + Size,
-	Time2 = Time + TimeDelta / 1000,
-	Performance2 = Performance#performance{ bytes = Bytes2, time = Time2,
-			rating = Rating2 = Bytes2 / (Time2 + 1), failures = 0, transfers = N + 1 },
-	Total2 = Total - Rating + Rating2,
-	may_be_rotate_peer_ports(Peer),
-	ets:insert(?MODULE, [{{peer, Peer}, Performance2}, {rating_total, Total2}]).
-
-get_or_init_performance(Peer) ->
+	
+	#performance{
+		average_bytes = AverageBytes,
+		total_bytes = TotalBytes,
+		average_latency = AverageLatency,
+		total_latency = TotalLatency,
+		average_success = AverageSuccess,
+		lifetime_rating = LifetimeRating,
+		current_rating = CurrentRating,
+		total_transfers = TotalTransfers
+	} = Performance,
+	TotalBytes2 = case DataSize of
+		undefined -> TotalBytes;
+		_ -> TotalBytes + DataSize
+	end,
+	%% AverageBytes is the average number of bytes transferred during the AverageLatency time
+	%% period. In order to approximate the impact of multiple concurrent requests we multiply
+	%% DataSize by the Concurrency value. We do this *only* when updating the AverageBytes
+	%% value so that it doesn't distort the TotalBytes.
+	AverageBytes2 = case DataSize of
+		undefined -> AverageBytes;
+		_ -> calculate_ema(AverageBytes, (DataSize * Concurrency), ?THROUGHPUT_ALPHA)
+	end,
+	TotalLatency2 = case LatencyMilliseconds of
```

Comment:
What do you think about computing the ratio (throughput) EMA directly? I am thinking about the sensitivity to occasional throughput spikes, like a Gbit/s transfer after the usual 100 Mbit/s - we should probably value consistent 200 MBit/s transfers higher than 100 MBit/s with very occasional mixes of 1 GBit/s transfers. The harmonic mean will be least sensitive to such spikes (`1 / (Alpha / NewThroughput + (1 - Alpa) / OldThroughput)` should be the harmonic EMA alternative). However, this might be very theoretical as in practice the performance of a single peer won't vary much and "occasional" spikes still mean paying for bandwidth for a potential attackerso we might just note it down for later.

Reply:
You mean instead of averaging data and latency separately, we would calculate the throughput for a single data point and then average that?

I'm fine with that change although I'm not entirely sure what the impact will be (I'll have to think through it some more). Would we do the same for the lifetime rating? 

I guess that one would become: (sum of all throughout) / TotalTransfers